test/e2e/aks_node_taints.go

```diff
@@ -0,0 +1,143 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type AKSNodeTaintsSpecInput struct {
+	Cluster       *clusterv1.Cluster
+	MachinePools  []*expv1.MachinePool
+	WaitForUpdate []interface{}
+}
+
+func AKSNodeTaintsSpec(ctx context.Context, inputGetter func() AKSNodeTaintsSpecInput) {
+	input := inputGetter()
+
+	settings, err := auth.GetSettingsFromEnvironment()
+	Expect(err).NotTo(HaveOccurred())
+	subscriptionID := settings.GetSubscriptionID()
+	auth, err := settings.GetAuthorizer()
+	Expect(err).NotTo(HaveOccurred())
+
+	managedclustersClient := containerservice.NewManagedClustersClient(subscriptionID)
+	managedclustersClient.Authorizer = auth
+
+	agentpoolsClient := containerservice.NewAgentPoolsClient(subscriptionID)
+	agentpoolsClient.Authorizer = auth
+
+	mgmtClient := bootstrapClusterProxy.GetClient()
+	Expect(mgmtClient).NotTo(BeNil())
+
+	infraControlPlane := &infrav1exp.AzureManagedControlPlane{}
+	err = mgmtClient.Get(ctx, client.ObjectKey{
+		Namespace: input.Cluster.Spec.ControlPlaneRef.Namespace,
+		Name:      input.Cluster.Spec.ControlPlaneRef.Name,
+	}, infraControlPlane)
+	Expect(err).NotTo(HaveOccurred())
+
+	var wg sync.WaitGroup
+
+	for _, mp := range input.MachinePools {
+		wg.Add(1)
+		go func(mp *expv1.MachinePool) {
+			defer GinkgoRecover()
+			defer wg.Done()
+
+			ammp := &infrav1exp.AzureManagedMachinePool{}
+			Expect(mgmtClient.Get(ctx, types.NamespacedName{
+				Namespace: mp.Spec.Template.Spec.InfrastructureRef.Namespace,
+				Name:      mp.Spec.Template.Spec.InfrastructureRef.Name,
+			}, ammp)).To(Succeed())
+			initialTaints := ammp.Spec.Taints
+
+			var expectedTaints infrav1exp.Taints
+			checkTaints := func(g Gomega) {
+				var expectedTaintStrs *[]string
+				if expectedTaints != nil {
+					expectedTaintStrs = new([]string)
+					*expectedTaintStrs = make([]string, 0, len(expectedTaints))
+					for _, taint := range expectedTaints {
+						*expectedTaintStrs = append(*expectedTaintStrs, fmt.Sprintf("%s=%s:%s", taint.Key, taint.Value, taint.Effect))
+					}
+				}
+
+				agentpool, err := agentpoolsClient.Get(ctx, infraControlPlane.Spec.ResourceGroupName, infraControlPlane.Name, *ammp.Spec.Name)
+				g.Expect(err).NotTo(HaveOccurred())
+				actualTaintStrs := agentpool.NodeTaints
+				if expectedTaintStrs == nil {
+					g.Expect(actualTaintStrs).To(BeNil())
+				} else {
+					g.Expect(actualTaintStrs).To(Equal(expectedTaintStrs))
+				}
+			}
+
+			Byf("Creating taints for machine pool %s", mp.Name)
+			expectedTaints = infrav1exp.Taints{
+				{
+					Effect: infrav1exp.TaintEffect(corev1.TaintEffectPreferNoSchedule),
+					Key:    "capz-e2e-1",
+					Value:  "test1",
+				},
+				{
+					Effect: infrav1exp.TaintEffect(corev1.TaintEffectPreferNoSchedule),
+					Key:    "capz-e2e-2",
+					Value:  "test2",
+				},
+			}
+			Expect(mgmtClient.Get(ctx, client.ObjectKeyFromObject(ammp), ammp)).To(Succeed())
+			ammp.Spec.Taints = expectedTaints
+			Expect(mgmtClient.Update(ctx, ammp)).To(Succeed())
```

Comment:
pending thoughts on https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3058 let's wrap the Get and Update methods inside a common retry loop

Reply:
done